circuit/primitives/src/xtx.rs

```diff
@@ -59,10 +65,458 @@ impl<
             delay_steps_at,
             result_status: None,
             total_reward,
+            full_side_effects,
         }
     }
 
     pub fn generate_xtx_id<T: frame_system::Config>(&self) -> XtxId<T> {
         SystemHashing::<T>::hash(Encode::encode(self).as_ref())
     }
+
+    pub fn is_completed(&self) -> bool {
+        for step in self.full_side_effects.iter() {
+            for full_side_effect in step.iter() {
+                if full_side_effect.confirmed.is_none() {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    // Complete the full side effect of Xtx by assigning confirmed side effect/
+    // This can only happen if the side effects is confirmed with respect to
+    // the execution steps.
+    //
+    // Return true if the input side effect successfully confirmed.
+    // Return false if no update has happened.
+    // Throw an error if detected the confirmation ruleset has been violated.
+    pub fn complete_side_effect<Hasher: sp_core::Hasher>(
+        &mut self,
+        confirmed: ConfirmedSideEffect<AccountId, BlockNumber, BalanceOf>,
+        input: SideEffect<AccountId, BlockNumber, BalanceOf>,
+    ) -> Result<bool, &'static str> {
+        let input_side_effect_id = input.generate_id::<Hasher>();
+
+        // Double check there are some side effects for that Xtx - should have been checked at API level tho already
+        if self.full_side_effects.is_empty() {
+            return Err("Xtx has no single side effect step to confirm");
+        }
+
+        let step_with_unconfirmed_steps: Option<usize>;
+
+        for (i, step) in self.full_side_effects.iter_mut().enumerate() {
+            // Double check there are some side effects for that Xtx - should have been checked at API level tho already
+            if step.is_empty() {
+                return Err("Xtx has an empty single step.");
+            }
+            for mut full_side_effect in step.iter_mut() {
+                if full_side_effect.confirmed.is_none() {
+                    step_with_unconfirmed_steps = Some(i);
+                    // Recalculate the ID for each input side effect and compare with the input one.
+                    // Check the current unconfirmed step before attempt to confirm the full side effect.
+                    return if full_side_effect.input.generate_id::<Hasher>() == input_side_effect_id
+                        && step_with_unconfirmed_steps == Some(i)
+                    {
+                        // We found the side effect to confirm from inside the unconfirmed step.
+                        full_side_effect.confirmed = Some(confirmed.clone());
+                        Ok(true)
```

Comment:
This result will probably depend on the contents of the ConfirmedSideEffect. The relayer might come back with some error and the SideEffect might be confirmed but unsuccessful.

Reply:
Content of ConfirmedSideEffect was checked before as a part of ConfirmExecution. 
Xtx only checks if the order of side effects is respected